### PR TITLE
Bugfix/2713 typedef transform inits

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary
Fixes compiler bug which produced c++ code that doesn't compile -
Added typedef to generated c++ code for `transform_inits` method

#### Intended Effect
generated c++ code compiles

#### How to Verify
run unit tests; added test model `src/test/test-models/good/parser-generator/test_transform_inits.stan`

#### Side Effects
N/A

#### Documentation
N/A

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by mitzimorris** [comment]
@seantalts if you have a chance, please review.

**Comment by seantalts** [comment]
@VMatthijs want to take a look? I don't have much experience in the current parser.

File: src/stan/lang/generator/generate_transform_inits_method.hpp [hunk 1]
```diff
@@ -41,6 +41,8 @@ namespace stan {
         << EOL;
       o << INDENT << "                     std::ostream* pstream__) const {"
         << EOL;
+      o << INDENT2 << "typedef double local_scalar_t__;"
```
Comment (bob-carpenter) [inline src/stan/lang/generator/generate_transform_inits_method.hpp]: I don't see what `local_scalar_t__` is doing---this is the only instance in the file.  

I wonder if it's sufficient given that parameters can be involved in the bounds, e.g.,
```
parameters {
  real alpha;
  real<lower = (alpha > 0 ? alpha : 0)> gamma;
}
```

**Comment by bob-carpenter** [comment]
I can deal with reviewing this one.

File: src/stan/lang/generator/generate_transform_inits_method.hpp (hunk 1 above)
Comment (mitzimorris) [reply src/stan/lang/generator/generate_transform_inits_method.hpp]: this change adds the above typedef as the first statement in the body of the `transform_inits` method.
if an expression used in the lower/upper bounds or location/scale (offset/multiplier) contains a parameter and the generated code requires a call to ` stan::math::promote_scalar`,  then the generated expression will be in terms of `local_scalar_t__` - e.g.:

```
            writer__.scalar_lb_unconstrain((logical_gt(alpha, 0) ? stan::math::promote_scalar<local_scalar_t__>(alpha) : stan::math::promote_scalar<local_scalar_t__>(0) ), gamma);
```

adding the typedef at the beginning of the `transform_inits` method makes this do the right thing.  note that the generated code for the constructor has exactly the same typedef as the first statement.

**Pull request merged**